### ISSUE TITLE
Add store weight/grad norm option

### ISSUE THREAD

**Issue by QuentinDuval** [facebookresearch/fairscale#696]
## 🚀 Feature

FSDP to offer the possibility to compute the norms of the weights and norms of the gradients on the fly, when the weights / gradients are available with an option like `compute_weight_norm=True` in the constructor and an API to retrieve those weights.

## Motivation

Following issue https://github.com/facebookresearch/fairscale/issues/644 and the attempt in solving it in PR https://github.com/facebookresearch/fairscale/pull/692, providing a view of the parameters in the view is too low level and does not offer the best user experience as there are plenty of ways to trip and fall (see PR for the limitations: https://github.com/facebookresearch/fairscale/pull/692).

Following a discussion with @min-xu-ai, the best is to go back at the uses cases motivating https://github.com/facebookresearch/fairscale/issues/644:

1. be able to easily compute the weight and gradient norms when `flatten_parameters=True` needed for LARC like optimisers
2. be able to have separate LR, regularisation for each parameters (for instance, regularise only bias and not weight or the other way around)

This issue proposes to solve item 1.

## Interested parties

CC: @min-xu-ai @myleott @prigoyal

**Comment by min-xu-ai** [comment]
Thanks for filing, @QuentinDuval!

I am thinking about a hook based approach where the hook callback is provided a list of the original "param tensors", like:

```
def forward_after_all_gather_hook(fsdp_module, named_param_tensors):
   results = {}
   for n, p in named_param_tensors:
        results[n] = p.norm()
```

On the backward, the hook could be:

```
def backward_after_full_grads(fsdp_module, n, param_tensor, grad_tensor):
   ...
```

The user however, must make sure he/she stores the results somewhere in a stable way since the full param or grad_tensors are going to be deleted due to sharding. Do you think the above APIs make sense?

**Comment by QuentinDuval** [comment]
> Thanks for filing, @QuentinDuval!
> 
> I am thinking about a hook based approach where the hook callback is provided a list of the original "param tensors", like:
> 
> ```
> def forward_after_all_gather_hook(fsdp_module, named_param_tensors):
>    results = {}
>    for n, p in named_param_tensors:
>         results[n] = p.norm()
> ```
> 
> On the backward, the hook could be:
> 
> ```
> def backward_after_full_grads(fsdp_module, n, param_tensor, grad_tensor):
>    ...
> ```
> 
> The user however, must make sure he/she stores the results somewhere in a stable way since the full param or grad_tensors are going to be deleted due to sharding. Do you think the above APIs make sense?

It would make sense indeed and would allow us to implement the LARC optimizer in a more optimized way 👍 

Some questions:
- Does the `forward_after_all_gather_hook` happen before or after the forward of the module is computed?
- Does the `backward_after_full_grads` argument named `n` represent the name of the parameter?

Note that in my case, for LARC use case, I would only need the backward hook.

In terms of naming, it must be made clear that these are not the same as the typical hooks which take the input (and outputs for post hooks) that pytorch has.

**Comment by min-xu-ai** [comment]
> Does the forward_after_all_gather_hook happen before or after the forward of the module is computed?

I was thinking before, but since you don't really need this, maybe we don't need to add it.

> Does the backward_after_full_grads argument named n represent the name of the parameter?

yes, I was thinking n is the name.

> In terms of naming, it must be made clear that these are not the same as the typical hooks which take the input (and outputs for post hooks) that pytorch has.

good point! we can make it something like "readonly_hook"?

**Comment by QuentinDuval** [comment]
> Thanks for filing, @QuentinDuval!
> 
> I am thinking about a hook based approach where the hook callback is provided a list of the original "param tensors", like:
> 
> ```
> def forward_after_all_gather_hook(fsdp_module, named_param_tensors):
>    results = {}
>    for n, p in named_param_tensors:
>         results[n] = p.norm()
> ```
> 
> On the backward, the hook could be:
> 
> ```
> def backward_after_full_grads(fsdp_module, n, param_tensor, grad_tensor):
>    ...
> ```
> 
> The user however, must make sure he/she stores the results somewhere in a stable way since the full param or grad_tensors are going to be deleted due to sharding. Do you think the above APIs make sense?

I was just realising that the backward hook would help computing LARC with `flatten_parameters=False` but not if `flatten_parameters=True`: there would still be a mapping problem from the information captured in the hook to the adaptative LR.

**Comment by min-xu-ai** [comment]
> but not if flatten_parameters=True: there would still be a mapping problem from the information captured in the hook to the adaptative LR.

True. Unless we waste some GPU memory by expand and concat p_norm and after_all_reduce_g_norm and shard drop the shards except the local shards. That way, each param element would have a correct p_norm and g_norm and padding is not an issue. Question is that do we have that much GPU memory to spare?

**Comment by QuentinDuval** [comment]
> > Does the forward_after_all_gather_hook happen before or after the forward of the module is computed?

> I was thinking before, but since you don't really need this, maybe we don't need to add it.

I like that 👍 

> good point! we can make it something like "readonly_hook"?

Indeed, the documentation should explain that return values are ignored. In terms of name, I was thinking about `on_gradient_computed_hook` which by the spelling somehow explain that this is just a listener, not an active part of the algorithm.

> True. Unless we waste some GPU memory by expand and concat p_norm and after_all_reduce_g_norm and shard drop the shards except the local shards. That way, each param element would have a correct p_norm and g_norm and padding is not an issue. Question is that do we have that much GPU memory to spare?

In any case, `flatten_parameters=False` or not, this would be a useful feature for us: @prigoyal and I discuted some time ago about the prospect of computing the weights & gradient norms before LARC is applied, so that things like gradient clipping can benefit from a shared computation.

This feature would make that much much easier!
